### PR TITLE
[tune] Cross-Node Recovery

### PR DESCRIPTION
**Pull request by richardliaw**
Augments trial restore to also check if the runner is at the same
location. If not, the checkpoint files are pushed onto the new location.

TODO:
 - [x] Test this on EC2
 - [ ] Autoscaler test (maybe add into stress tests?)
    - Option 1: Have a helper fn that blocks until the ray cluster is the expected size, then submit tune job, down scale cluster, and then make sure the exit code of the submit job is 0.

We don't have any tests for this right now...

### CODE REVIEW

File: python/ray/tune/ray_trial_executor.py [hunk 1]
```diff
@@ -327,9 +327,11 @@ def restore(self, trial, checkpoint=None):
                 assert type(value) != Checkpoint, type(value)
                 ray.get(trial.runner.restore_from_object.remote(value))
             else:
+                current_runner_ip = ray.get(trial.runner.current_ip.remote())
+                trial.update_location(current_runner_ip)
```
Comment (richardliaw) [inline python/ray/tune/ray_trial_executor.py]:
```suggestion
                trial.update_logger_location(current_runner_ip)
```

File: python/ray/tune/logger.py [hunk 2]
```diff
@@ -122,6 +123,12 @@ def flush(self):
         self._log_syncer.sync_now(force=True)
         self._log_syncer.wait()
 
+    def update_location(self, worker_ip):
+        """Sends the current log directory to the remote node."""
+        if worker_ip != self._log_syncer.worker_ip:
```
Comment (ericl) [inline python/ray/tune/logger.py]: Will this crash on non-autoscaler clusters?

Also, log a warning if sync is happening?

File: python/ray/tune/logger.py [hunk 3]
```diff
@@ -122,6 +123,12 @@ def flush(self):
         self._log_syncer.sync_now(force=True)
         self._log_syncer.wait()
 
+    def update_location(self, worker_ip):
```
Comment (ericl) [inline python/ray/tune/logger.py]: update_location() sounds more harmless than this is, maybe sync_results_to_new_location()?

File: python/ray/tune/logger.py (hunk 2 above)
Comment (richardliaw) [reply python/ray/tune/logger.py]: Done; and no, won't crash (1sync_to_worker_if_possible1 catches it)

File: python/ray/tune/ray_trial_executor.py [hunk 4]
```diff
@@ -312,7 +318,10 @@ def save(self, trial, storage=Checkpoint.DISK):
         return trial._checkpoint.value
 
     def restore(self, trial, checkpoint=None):
-        """Restores training state from a given model checkpoint."""
+        """Restores training state from a given model checkpoint.
+
+        This will also sync the trial results to a new location
+        if restoring on a different node."""
```
Comment (richardliaw) [inline python/ray/tune/ray_trial_executor.py]:
```suggestion
        if restoring on a different node.
        """
```

File: python/ray/tune/ray_trial_executor.py [hunk 5]
```diff
@@ -327,11 +336,13 @@ def restore(self, trial, checkpoint=None):
                 assert type(value) != Checkpoint, type(value)
                 ray.get(trial.runner.restore_from_object.remote(value))
             else:
+                worker_ip = ray.get(trial.runner.current_ip.remote())
+                trial.sync_logger_to_new_location(worker_ip)
                 ray.get(trial.runner.restore.remote(value))
-            trial.last_result = checkpoint.last_result
```
Comment (richardliaw) [inline python/ray/tune/ray_trial_executor.py]:
```suggestion
```

**Comment by richardliaw** [comment]
This is ready for another review.

**Review by ericl — APPROVED** [review]

**Pull request merged**